### PR TITLE
add portable error handling

### PR DESCRIPTION
**Pull request by Yurlungur**
## PR Summary

In [the Vinet PR](https://github.com/lanl/singularity-eos/pull/202) by @aematts in singularity-eos some issues with the current error handling in singularity-eos were noted. I believe a performance-portable error handling treatment should be developed and properly lives in `ports-of-call`. I introduce it here. In particular, I define several flavors of macro:
- `PORTABLE_REQUIRE` which is an assert statement with file and line numbers
- `PORTABLE_REQUIRE_THROWS` as above but throws an exception
- `PORTABLE_ABORT` which is a fail statement with file and line numbers, resolves to backend aborts appropriate
- `PORTABLE_THROW` resolves to throwing an exception
- `PORTABLE_THROW_IFEXCEPT` resolves to either an exception or an abort depending on portability strategy
- `PORTABLE_WARN` which is a warning with file and line numbers 

All the above are no-ops by default, but are enabled for debug builds. I also provide `PORTABLE_ALWAYS_*` macros which always work.

The idea here is by default error handling should be silent unless you're debugging. But if the developer knows they want a message or a failure, they can insert one. For the Vinet PR, initialization errors should probably be handled with `PORTABLE_THROW_IFEXCEPT` for example.

## PR Checklist

- [x] Any changes to code are appropriately documented.
- [x] Code is formatted.
- [ ] Install test passes.
- [ ] Docs build.
- [ ] If preparing for a new release, update the version in cmake.

### CODE REVIEW

File: ports-of-call/portable_errors.hpp
Comment (mauneyc-LANL) [inline ports-of-call/portable_errors.hpp:57]: probably `#ifdef`

File: doc/sphinx/src/using.rst
Comment (jhp-lanl) [inline doc/sphinx/src/using.rst:88]: Given that many of these _do not_ work on device, what then is the definition of "portable"?

File: ports-of-call/portable_errors.hpp
Comment (jhp-lanl) [inline ports-of-call/portable_errors.hpp:79]: Shouldn't the `KOKKOS_INLINE_FUNCTION` be behind the `ifdef`?

**Review by jhp-lanl — COMMENTED** [review]
This will be very helpful I think. But rather than some of these just not working on device, should they instead do nothing? My issue is that if a "portable" implementation doesn't work on device, it begs the question as to whether it's truly portable anymore.

File: ports-of-call/portable_errors.hpp
Comment (jhp-lanl) [reply ports-of-call/portable_errors.hpp:79]: Or should it use `PORTABLE_FUNCTION`?

File: ports-of-call/portable_errors.hpp
Comment (Yurlungur) [reply ports-of-call/portable_errors.hpp:57]: thanks for the catch

**Comment by Yurlungur** [comment]
> But rather than some of these just not working on device, should they instead do nothing? My issue is that if a "portable" implementation doesn't work on device, it begs the question as to whether it's truly portable anymore.

That's a fair point. This was written this way to support exception handling on host, when that's desirable. For example, you can imagine some initialization function that uses C++ std exception logic. Basically I wanted to proved that, also for testing. For example the catch machinery Anne wants use for Vinet catches exceptions.

File: doc/sphinx/src/using.rst
Comment (Yurlungur) [reply doc/sphinx/src/using.rst:92]: Yeah that makes sense. I'll change the name.

File: doc/sphinx/src/using.rst
Comment (Yurlungur) [reply doc/sphinx/src/using.rst:88]: In this case I'm just namespacing, because the macros can't live in a `PortsOfCall` namespace, so really `PORTABLE` here is `PORTS_OF_CALL`

I think it makes sense to provide a unified interface though with the truly portable versions and the not-so-portable versions that use exception handling. The issue is exception handling simply does not work on device, but we may still want to provide it. Here by exception handling I mean specifically the machinery provided by `stdexcept` which provides things like
```C++
try {
  // some code
  if (condition) {
    PORTABLE_THROW(message)
  } catch (const std::runtime_error &error) {
    // some code that recovers from the error gracefully
  }
}
```
what `ports-of-call` is providing on top of that is a filename and line number, as well as `THROW_OR_ABORT` which gracefully switches between the exception model and the "print an error and die" model.

File: ports-of-call/portable_errors.hpp
Comment (Yurlungur) [reply ports-of-call/portable_errors.hpp:79]: It should just be `PORTABLE_INLINE_FUNCTION` my bad.

File: doc/sphinx/src/using.rst
Comment (Yurlungur) [reply doc/sphinx/src/using.rst:88]: I suppose we could not provide `REQUIRE_THROWS` or `THROWS` and only provide `THROW_OR_ABORT`. But I think sometimes we actually do want to just throw.

Comment (chadmeyer) [reply doc/sphinx/src/using.rst:88]: > This will be very helpful I think. But rather than some of these just not working on device, should they instead do nothing? My issue is that if a "portable" implementation doesn't work on device, it begs the question as to whether it's truly portable anymore.

I recognize that having a *universal* error handling mechanism is probably a pipe dream, but is it not possible to have something that could yet function on device and do something?  I seem to think that we've had things that can abort a kernel and then set some error state (which may not be seen until the next kernel launch) but might be limited to CUDA or something like that (which, to be clear, we should support).  I'm certainly no expert on error handling!  I think we should be careful about putting together something that could easily be snuck in to some inappropriate parts of library codes that would then make for some downstream headaches; at the very least if the macros have to expand to nothing on device that would be better than having them expand to something with undefined behavior.

Comment (Yurlungur) [reply doc/sphinx/src/using.rst:88]: It is possible to have something that throws an error on host and aborts on device. I could remove the functions that only work on host if that's how people feel, and leave things that are all truly portable. The only disadvantage of this is that we will be excluding any use of C++ exceptions if anyone cares about that. But I suppose if someone wants to use that it can be on them.

Comment (chadmeyer) [reply doc/sphinx/src/using.rst:88]: My main concern is the case where a downstream library puts in this kind of functionality to ensure certain behavior, but in reality this is device code (at least sometimes) and the behavior becomes either undefined or wrong/unexpected.  I can imagine it being a very hard bug to track down if there is some kind of silent-ish abort of a device-side constructor or something similar but the host side is never the wiser.  

Maybe I'm being a little too paranoid about this here?

Comment (Yurlungur) [reply doc/sphinx/src/using.rst:88]: so I *think* the functionality that doesn't work on device will fail to compile. But I should double check that. We *can* simply not provide the functionality that might fail. The reason I'm reluctant to do that is I feel that there are times when C++ exceptions provide a useful ability on host.

The aborts that *do* run on device will all run on host or device and will be loud.

File: ports-of-call/portable_errors.hpp
Comment (jhp-lanl) [inline ports-of-call/portable_errors.hpp:153]: Why not template on the type of `msg` rather than overloading on explicit types?

File: doc/sphinx/src/using.rst
Comment (jhp-lanl) [reply doc/sphinx/src/using.rst:88]: > In this case I'm just namespacing, because the macros can't live in a `PortsOfCall` namespace, so really `PORTABLE` here is `PORTS_OF_CALL`
> 
> I think it makes sense to provide a unified interface though with the truly portable versions and the not-so-portable versions that use exception handling. The issue is exception handling simply does not work on device, but we may still want to provide it. Here by exception handling I mean specifically the machinery provided by `stdexcept` which provides things like
> 
> ```c++
> try {
>   // some code
>   if (condition) {
>     PORTABLE_THROW(message)
>   } catch (const std::runtime_error &error) {
>     // some code that recovers from the error gracefully
>   }
> }
> ```
> 
> what `ports-of-call` is providing on top of that is a filename and line number, as well as `THROW_OR_ABORT` which gracefully switches between the exception model and the "print an error and die" model.

But what is the true _utililty_ of `PORTABLE_THROW`? What if I modify the above code (I assume it should be try/catch rather than try{if/catch}, but my C++ knowledge is limited) to be
```c++
try {
   // some code
   if (condition) {
      const char* const message = PORTABLE_ERR_MSG(message)
      std::runtime_error(msg)
   }
} catch (const std::runtime_error &error) {
   // some code that recovers from the error gracefully
}
```
If the utility is in the creation of the error message, should that then be the functionality that is exposed?

Comment (Yurlungur) [reply doc/sphinx/src/using.rst:88]: Hmm. Yeah that's a good point. I like the idea of exposing the error message generation and removing the throw machinery... with the exception of the macro that either throws or aborts depending.

Comment (Yurlungur) [reply doc/sphinx/src/using.rst]: Ok I've made the change. Here's how it works now:
- `PORTABLE_THROW` is gone.
- `PORTABLE_THROW_OR_ABORT` still exists and is the preferred way to perform checks that you would like to throw an exception but want to work on device.
- I have added `PORTABLE_ERROR_MESSAGE` . It can't work as cleanly as your suggested code above @jhp-lanl because that kind of assignment leaks memory, so you have to fill an array that is passed in as an argument. It also has to be a `char*` to work on device, which means the user is responsible for allocating enough memory to support their string on device without a segfault. But I think that's fine.

Comment (jhp-lanl) [reply doc/sphinx/src/using.rst]: I was thinking about this yesterday but forgot to chime in. Would a compile time error message length solve the memory leak issue? Could the length default to something and be set via CMake?

BUUUUUT, I don't feel strongly. Simply printing something and then raising an exception is also fine with me.

**Review by jhp-lanl — APPROVED** [review]
Provisionally approving since I think this addresses my comments. I'd like @dholladay00 or @chadmeyer to give final approval though

File: doc/sphinx/src/using.rst
Comment (Yurlungur) [reply doc/sphinx/src/using.rst]: > I was thinking about this yesterday but forgot to chime in. Would a compile time error message length solve the memory leak issue? Could the length default to something and be set via CMake?

That would allow us to fix length ahead of time, which would avoid the segfault thing, but it wouldn't allow us to return a pointer. We'd still have to fill an array that's passed to us. I think the current design is fine, though I'm not opposed to a compile-time string length if people feel strongly.

**Comment by Yurlungur** [comment]
Yep I'll wait to hear from @chadmeyer and @dholladay00 before merging.

**Comment by Yurlungur** [comment]
@dholladay00 and @chadmeyer what are your thoughts here?

**Comment by mauneyc-LANL** [comment]
Some documentation in the code wouldn't hurt, though I leave it to your discretion.

**Comment by Yurlungur** [comment]
> Some documentation in the code wouldn't hurt, though I leave it to your discretion.

I don't mind adding more documentation. I didn't add any here because I didn't see what would be a useful addition in this case. What would you find helpful @mauneyc-LANL ?

**Comment by mauneyc-LANL** [comment]
> > Some documentation in the code wouldn't hurt, though I leave it to your discretion.
> 
> I don't mind adding more documentation. I didn't add any here because I didn't see what would be a useful addition in this case. What would you find helpful @mauneyc-LANL ?

Really just a bit of comments to help "paragraph" the code a little and make it easier to parse.

```c++
/***
Used when conditions are encountered where execution could not complete. 
Takes an `input_message`, `filename`, and `line_no`  describing the error, and 
wraps it into formatted output `output_message`.
***/
PORTABLE_INLINE_FUNCTION
void error_msg(const char *const input_message, const char *const filename,
               int const linenumber, char *output_message) {
  std::sprintf(output_message,
               "### ERROR\n  Message:     %s\n  File:        %s\n  Line number: %i\n",
               input_message, filename, linenumber);
}
```
(probably something less illiterate)

File: ports-of-call/portable_errors.hpp
Comment (mauneyc-LANL) [inline ports-of-call/portable_errors.hpp:179]: If I follow, you're missing the `output_message` parameter here.

**Comment by Yurlungur** [comment]
> > > Some documentation in the code wouldn't hurt, though I leave it to your discretion.
> > 
> > 
> > I don't mind adding more documentation. I didn't add any here because I didn't see what would be a useful addition in this case. What would you find helpful @mauneyc-LANL ?
> 
> Really just a bit of comments to help "paragraph" the code a little and make it easier to parse.
> 
> ```c++
> /***
> Used when conditions are encountered where execution could not complete. 
> Takes an `input_message`, `filename`, and `line_no`  describing the error, and 
> wraps it into formatted output `output_message`.
> ***/
> PORTABLE_INLINE_FUNCTION
> void error_msg(const char *const input_message, const char *const filename,
>                int const linenumber, char *output_message) {
>   std::sprintf(output_message,
>                "### ERROR\n  Message:     %s\n  File:        %s\n  Line number: %i\n",
>                input_message, filename, linenumber);
> }
> ```
> 
> (probably something less illiterate)

Sure I'll add a little bit.

**Comment by Yurlungur** [comment]
@mauneyc-LANL take a look now. Does this help? I tired to describe the intent of the design as well... e.g., why there are both macros and functions.

**Comment by mauneyc-LANL** [comment]
> @mauneyc-LANL take a look now. Does this help? I tried to describe the intent of the design as well... e.g., why there are both macros and functions.

looks good

**Comment by jhp-lanl** [comment]
@Yurlungur is this ready now?

File: ports-of-call/portable_errors.hpp
Comment (dholladay00) [inline ports-of-call/portable_errors.hpp:117]: won't this double print the message? same for below

Comment (Yurlungur) [reply ports-of-call/portable_errors.hpp:117]: yes it does---personally I think this is fine, but if it bothers you I can tweak to avoid that

**Comment by Yurlungur** [comment]
> @Yurlungur is this ready now?

as soon as I get OK from @dholladay00 and/or @chadmeyer

File: ports-of-call/portable_errors.hpp
Comment (dholladay00) [reply ports-of-call/portable_errors.hpp:117]: If this happens to get used at a low level, it could get very annoying (ie **twice** per material, per iteration (if pte solve), per cell, per rank, per time step). Could you just pass in an empty string?

Comment (Yurlungur) [reply ports-of-call/portable_errors.hpp]: Yeah I'll do that.

**Comment by Yurlungur** [comment]
@dholladay00 how's it look now?

**Review by dholladay00 — APPROVED** [review]

**Pull request merged**
